### DISCUSSION TITLE
Issue with Adding Components "Dynamically" - Grouping Components Based on Property Value

### DISCUSSION THREAD

**Discussion by MsnVt** [wixtoolset/issues#7464] [Questions]
Which version of WiX are you building with?
3.11

Which version of Visual Studio are you building with (if any)?
2022

Which version of .NET are you building with?
4.8.09032

Describe the problem and the steps to reproduce it.
I'm currently facing an issue while working with the WiX toolset, specifically related to adding components to my installer. I am hoping to get some guidance or suggestions from the community to resolve this challenge. (A fair warning, I'm new to using the WiX toolset and with working with installers altogether, so I ask that you bare with me if I make any mistakes)

Here's the problem: I would like to use the heat tool to generate components and source files. However, I want the installer to install a specific group of components based on a property-value that the user selects from a dialog during installation. Instead of referencing or creating a component for each individual file inside a given directory, I want to install components as a "group" based on certain conditions.

Ideally, I'd like to define several directories, each with their own condition. If the condition evaluates to true, I want the installer to install the corresponding directory and its components. I believe this approach would provide flexibility and streamline the installer configuration process.

Has anyone encountered a similar scenario or implemented a similar functionality using the WiX toolset? I would greatly appreciate any insights or advice on how to achieve this goal.

Are there any built-in features within the WiX toolset that can help with grouping components based on property values? Or perhaps there are alternative approaches that have been successful in achieving similar results?

I have explored the documentation and various resources available online, but I haven't found a straightforward solution for this specific use case. Any guidance or examples would be highly appreciated.

Thank you in advance for your time and assistance. I look forward to hearing your thoughts and experiences on this matter.

As you can see below there is a directory being made and a couple components being added as well. Ideally, I'd want to limit what I'm explicitly creating to only the respective directories(d1, d2, etc…) under the premise that I'd be able to "somehow" install the files within that directory by referencing my heat generated components/source-code.
`
							
						<Directory Id="DIRECTORY1" Name="D1”>
							<Component Id="Coreresourcesdll_D1_comp" Guid="{12345678-1234-1234-1234-123456789123}" Transitive="yes">
								<Condition><![CDATA[DIRECTORY1=SELECTION Or DIRECTORY1="True"]]></Condition>
								<File Id="Coreresourcesdll_D1" KeyPath="yes" Source="$(var.HarvestPath)D1\Core.resources.dll" />
							</Component>
							<Component Id="Coreresourcesdll_D2_comp" Guid="{12345678-1234-1234-1234-123456789123}" Transitive="yes">
								<Condition><![DIRECTORY2=SELECTION Or DIRECTORY2="True"]]></Condition>
								<File Id="Coreresourcesdll_D2" KeyPath="yes" Source="$(var.HarvestPath)D2/Core.resources.dll" />
							</Component>
						
							...
							`
							
							

To put it short, I want to be able to load in a directory and all of its file contents without having to explicitly create a component for each file within the given directory. I then want to be able to control which of those directories will be targeted for installation, using conditions.

**Comment by BMurri** [comment]
Some things to think about:

In Windows Installer (aka MSI) Components are the grouping of resources in each "atomic" unit of deployment. If you don't want to go down the very large "rabbit hole" quite yet of understanding all the implications of that, stick hard and fast to the rule of "one file per component".

In Windows Installer (aka MSI) Features are the grouping of components in each "selectable" unit of deployment. Adding conditions to components (especially when those conditions may be intended to change, such as customer configuration selections) usually leads to unexpected or unintended consequences. Reserve conditions on components only for "permanent" environmental issues. Add conditions to Features to toggle customer configurations/preferences.

There should be almost no case where it makes sense to include a GUID in any WiX source code, and that includes the output of the harvester (aka `heat`). In WiX v4, remove all @Guid everywhere (and don't let heat create them), unless it's setting all of them to `"*"`.

A defined directory will, by default, only be created if there are files selected for installation to go into them.
